### ISSUE TITLE
strange ArrayIndexOutOfBoundsException during deserialization, depending on type parameters

### ISSUE THREAD

**Issue by reuschling** [jdereg/json-io#33]
I have a strange error while deserializing one of my data structures. The structure is too complex to post it here, but I managed it to create a small snippet to reproduce the problem:

```java
import com.cedarsoftware.util.io.JsonReader;
import com.cedarsoftware.util.io.JsonWriter;


public class FeldWaldWiese
{
    static public class Test<K, V>
    {
        // WORKING:
//        protected Test2 m_internalHashMap = new Test2();
        // NOT WORKING:
        protected Test2<K, V> m_internalObject = new Test2<>();
    }

    static public class Test2<K, V>
    {
        //strange: if I leave just one of the members, it is working.
        //Maybe the member count is relevant. It seems to be independent from the members type
        int i;
        int y;
        String check;
    }

    public static void main(String[] args) throws Exception
    {
        Test item = new Test();

        JsonReader.jsonToJava(JsonWriter.objectToJson(item));
    }
}
```
the snippet generates following exception:
```
Exception in thread "main" java.io.IOException: ArrayIndexOutOfBoundsException setting field 'm_internalObject' on target: org.dynaq.FeldWaldWiese$Test@4e5a5622 with value: {i=0, y=0, check=null}
Last read: {"@type":"org.dynaq.FeldWaldWiese$Test","m_internalObject":{"i":0,"y":0,"check":null}}
line: 1, col: 87
	at com.cedarsoftware.util.io.JsonReader.error(JsonReader.java:3238)
	at com.cedarsoftware.util.io.JsonReader.assignField(JsonReader.java:1894)
	at com.cedarsoftware.util.io.JsonReader.traverseFields(JsonReader.java:1762)
	at com.cedarsoftware.util.io.JsonReader.convertMapsToObjects(JsonReader.java:1294)
	at com.cedarsoftware.util.io.JsonReader.convertParsedMapsToJava(JsonReader.java:1233)
	at com.cedarsoftware.util.io.JsonReader.readObject(JsonReader.java:1193)
	at com.cedarsoftware.util.io.JsonReader.jsonToJava(JsonReader.java:1085)
	at org.dynaq.FeldWaldWiese.main(FeldWaldWiese.java:32)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 2
	at com.cedarsoftware.util.io.JsonReader.markUntypedObjects(JsonReader.java:1991)
	at com.cedarsoftware.util.io.JsonReader.assignField(JsonReader.java:1792)
	... 6 more
```
If I leave the type parameters from the member in the container class 'Test', no exception will be thrown. If I leave one of the parameters inside the class 'Test2', also no exception is thrown. It seems to me that the problem starts if more than 2 member variables are in the second class.

The snippet was executed with Oracle  Java JDK 1.7.0_60.

**Comment by jdereg** [comment]
What version of json-io are you running?  Try 2.9.0 and see if you still have the issue.

**Comment by reuschling** [comment]
You are right and I'm sorry, I thinked I had 2.9.0, but accidently Maven prefered a conflicting 2.7.2 for reasons I don't know.
Nevertheless, with 2.9.0, I fall into another failure. I was able to create a test snippet for it also:

```java
    // NOT WORKING (with type parameters)
    static public class Test
    {
        protected Test2<String, Object> internalMember = new Test2<>();
    }

    static public class Test2<K, V>
    {
        protected Map<K, Collection<? extends V>> hsOne = new HashMap<>();;

        // Exception during deserialisation while setting this member
        protected Class<?> two = LinkedList.class;
    }

    // WORKING (without type parameters)
//    static public class Test
//    {
//        protected Test2 internalMember = new Test2();
//    }
//
//    static public class Test2
//    {
//        protected Map hsOne = new HashMap();
//
//        protected Class two = LinkedList.class;
//    }



    public static void main(String[] args) throws Exception
    {
        Test container = new Test();

        Test container2 = new Test();

        LinkedList<Test> llList = new LinkedList<>();
        llList.add(container2);

        container.internalMember.hsOne.put("key", llList);


        JsonReader.jsonToJava(JsonWriter.objectToJson(container));
    }
```
Deserialisation also fails depending on having type parameters or not. The snippet throws following exception:

```
Exception in thread "main" java.io.IOException: IllegalArgumentException setting field 'two' on target: org.dynaq.FeldWaldWiese$Test2@7466a008 with value: {value=java.util.LinkedList}
Last read: :"java.util.LinkedList","@items":[{"@type":"org.dynaq.FeldWaldWiese$Test","internalMember":{"hsOne":{"@type":"java.util.HashMap"},"two":{"@id":1,"value":"java.util.LinkedList"}}}]}},"two":{"@ref":1}}}
line: 1, col: 309
	at com.cedarsoftware.util.io.JsonReader.error(JsonReader.java:3440)
	at com.cedarsoftware.util.io.JsonReader.assignField(JsonReader.java:1983)
	at com.cedarsoftware.util.io.JsonReader.traverseFields(JsonReader.java:1852)
	at com.cedarsoftware.util.io.JsonReader.convertMapsToObjects(JsonReader.java:1379)
	at com.cedarsoftware.util.io.JsonReader.traverseCollection(JsonReader.java:1670)
	at com.cedarsoftware.util.io.JsonReader.convertMapsToObjects(JsonReader.java:1371)
	at com.cedarsoftware.util.io.JsonReader.convertParsedMapsToJava(JsonReader.java:1318)
	at com.cedarsoftware.util.io.JsonReader.readObject(JsonReader.java:1278)
	at com.cedarsoftware.util.io.JsonReader.jsonToJava(JsonReader.java:1169)
	at org.dynaq.FeldWaldWiese.main(FeldWaldWiese.java:59)
Caused by: java.lang.IllegalArgumentException: Can not set java.lang.Class field org.dynaq.FeldWaldWiese$Test2.two to java.lang.String
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:164)
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:168)
	at sun.reflect.UnsafeObjectFieldAccessorImpl.set(UnsafeObjectFieldAccessorImpl.java:81)
	at java.lang.reflect.Field.set(Field.java:741)
	at com.cedarsoftware.util.io.JsonReader.assignField(JsonReader.java:1908)
	... 8 more
```

**Comment by jdereg** [comment]
Thank you for the updated information and the test case.  I fixed this issue and released a 2.9.1 of json-io with this fix.  I added your test and another test that illustrated the bug more clearly (to me).  See TestTemplateField class in the src/groovy package.

**Comment by reuschling** [comment]
It works like a charme - thanks a lot
